colcon_core/shell/__init__.py

```diff
@@ -354,3 +356,128 @@ def get_colcon_prefix_path(*, skip):
+            'built:' +
+            ''.join('\n- %s' % name for name in env_packages.keys()) +
+            '\nThey are being used from the following locations instead:' +
+            ''.join('\n- %s' % path for path in env_packages.values()) +
+            '\nTo suppress this warning ignore these packages in the ' +
+            'workspace:\n--packages-ignore ' + ' '.join(env_packages.keys()))
+
+    # raise error in case any dependencies are not matched
+    if missing:
+        raise RuntimeError(
+            'Failed to find the following files:' +
+            ''.join('\n- %s' % path for path in missing.values()) +
+            '\nCheck that the following packages have been built:' +
+            ''.join('\n- %s' % name for name in missing.keys()))
+
+
+def find_installed_packages_in_environment():
+    """
+    Find packages under the COLCON_PREFIX_PATH.
+
+    For each prefix path the package index is being read and the first time a
+    package is being found its install prefix is being added to the result.
+
+    :returns: The mapping from a package name to the prefix path
+    :rtype: OrderedDict
+    """
+    packages = OrderedDict()
+    for prefix_path in get_colcon_prefix_path():
+        prefix_path = Path(prefix_path)
+        pkgs = find_installed_packages(prefix_path)
+        if pkgs is None:
+            logger.debug(
+                "Ignoring prefix path '{prefix_path}'".format_map(locals()))
+            continue
+        for pkg_name in sorted(pkgs.keys()):
+            # ignore packages with the same name in "lower" prefix path
+            if pkg_name in packages:
+                continue
+            packages[pkg_name] = pkgs[pkg_name]
+    return packages
+
+
+def find_installed_packages(install_base):
+    """
+    Find install packages under the install base path.
+
+    The path must contain a marker file with the install layout.
+    Based on the install layout the packages are discovered i different
+    locations.
+
+    :param Path install_base: The base path to find installed packages in
+    :returns: The mapping from a package name to the prefix path, None if the
+      path doesn't exist or doesn't contain a marker file with a valid install
+      layout
+    :rtype: OrderedDict or None
+    """
+    marker_file = install_base / '.colcon_install_layout'
+    if not marker_file.is_file():
+        return None
+    install_layout = marker_file.read_text().rstrip()
+    if install_layout not in ('isolated', 'merged'):
+        return None
+
+    packages = {}
+    if install_layout == 'isolated':
+        # for each subdirectory look for the package specific file
+        for p in install_base.iterdir():
+            if not p.is_dir():
+                continue
+            if p.name.startswith('.'):
+                continue
+            p = p / get_relative_package_index_path() / p.name
+            if p.is_file():
+                packages[p.name] = p
```

Comment:
storing a file in the dictionary here seems contrary to the claim at https://github.com/colcon/colcon-core/pull/80/files#diff-4f967364532f7970bf2143e67da82f54R452 that it is a mapping from package names to prefix paths?

Reply:
`p` is not a file. It is a `pathlib.Path` instance.

Reply:
from [the line above](https://github.com/colcon/colcon-core/pull/80#discussion-diff-197555946R473) though that path instance is pointing to a file not a directory

Reply:
This has also been addressed in ab37b48dd6b7e626ee155056289ba771f5dd12a5. to use the initial `p` and not the "updated" `p`.